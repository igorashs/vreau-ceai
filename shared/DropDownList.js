import styled, { css } from 'styled-components';
import ArrowSvg from 'assets/icons/arrow.svg';
import { Button } from './Button';
import { useState } from 'react';

const itemStyle = css`
  padding: calc(var(--baseline) / 4) 0;
  border-bottom: 1px solid var(--layout);
`;

const Header = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;

  ${itemStyle}

  p {
    font-weight: 500;
  }

  button {
    transform: rotate(-90deg);
  }

  ${({ hide }) => hide && 'button {transform: rotate(90deg);}'}
`;

const List = styled.ul`
  display: flex;
  flex-direction: column;

  li {
    ${itemStyle}
  }

  ${({ hide }) => hide && 'display: none;'}
`;

export function DropDownList({ label, children }) {
  const [hide, setHide] = useState(false);

  return (
    <div>
      <Header hide={hide}>
        <p>{label}</p>
        <Button icon onClick={() => setHide((e) => !e)} noPadding>
          <ArrowSvg />
        </Button>
      </Header>
      <List hide={hide}>{children}</List>
    </div>
  );
}
