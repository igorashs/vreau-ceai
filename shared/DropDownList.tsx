import styled, { css } from 'styled-components';
import ArrowSvg from 'assets/icons/arrow.svg';
import { useState } from 'react';
import Button from './Button';

const itemStyle = css`
  padding: calc(var(--baseline) / 4) 0;
  border-bottom: 1px solid var(--layout);
`;

type HideStyledProps = {
  hide?: boolean;
};

const Header = styled.div<HideStyledProps>`
  display: flex;
  align-items: center;
  justify-content: space-between;

  ${itemStyle}

  p {
    font-weight: 500;
  }

  button {
    transform: rotate(-90deg);
  }

  ${({ hide }) => hide && 'button {transform: rotate(90deg);}'}
`;

const List = styled.ul<HideStyledProps>`
  display: grid;
  column-gap: var(--baseline);
  grid-template-columns: repeat(auto-fit, minmax(224px, 1fr));

  li {
    ${itemStyle}
  }

  ${({ hide }) => hide && 'display: none;'}
`;

type DropDownListProps = {
  label: string;
};

const DropDownList = ({
  label,
  children,
}: React.PropsWithChildren<DropDownListProps>) => {
  const [hide, setHide] = useState(false);

  return (
    <div>
      <Header hide={hide}>
        <p>{label}</p>
        <Button icon onClick={() => setHide((e) => !e)} noPadding>
          <ArrowSvg />
        </Button>
      </Header>
      <List hide={hide}>{children}</List>
    </div>
  );
};

export default DropDownList;
