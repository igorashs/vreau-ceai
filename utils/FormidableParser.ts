import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
