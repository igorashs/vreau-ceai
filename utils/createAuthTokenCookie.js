import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
