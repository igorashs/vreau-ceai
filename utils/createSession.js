import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
