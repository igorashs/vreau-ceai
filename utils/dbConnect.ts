import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
