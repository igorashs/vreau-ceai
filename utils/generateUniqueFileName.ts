import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
