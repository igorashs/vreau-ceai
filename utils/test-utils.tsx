import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
