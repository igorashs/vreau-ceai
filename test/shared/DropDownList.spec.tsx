import { render, fireEvent } from '@/utils/test-utils';
import DropDownList from '@/shared/DropDownList';

describe('DropDownList', () => {
  it('should render correctly', () => {
    const label = 'my list';

    const { container, getByText } = render(
      <DropDownList label={label}>
        <li>item1</li>
        <li>item2</li>
      </DropDownList>,
    );

    expect(getByText(label)).toBeVisible();
    expect(getByText('item1')).toBeVisible();
    expect(getByText('item2')).toBeVisible();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('.onButtonClick', () => {
    it('toggles list', () => {
      const label = 'my list';

      const { getByText, getByRole } = render(
        <DropDownList label={label}>
          <li>item1</li>
          <li>item2</li>
        </DropDownList>,
      );

      fireEvent.click(getByRole('button'));
      const listItem = getByText('item1');

      expect(listItem).not.toBeVisible();
      expect(listItem.parentElement).not.toBeVisible();
      expect(listItem.parentElement).toHaveStyleRule('display', 'none');
      expect(getByRole('button').parentElement).toHaveStyleRule(
        'transform',
        'rotate(90deg)',
        {
          modifier: 'button',
        },
      );

      fireEvent.click(getByRole('button'));
      expect(listItem).toBeVisible();
    });
  });
});
