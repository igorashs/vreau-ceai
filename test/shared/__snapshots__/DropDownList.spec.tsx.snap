// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownList should render correctly 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: 4px;
  box-shadow: none;
  padding: 0;
}

.c1:hover {
  background-color: var(--accent);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: calc(var(--baseline) / 4) 0;
  border-bottom: 1px solid var(--layout);
}

.c0 p {
  font-weight: 500;
}

.c0 button {
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.c2 {
  display: grid;
  -webkit-column-gap: var(--baseline);
  column-gap: var(--baseline);
  grid-template-columns: repeat(auto-fit,minmax(224px,1fr));
}

.c2 li {
  padding: calc(var(--baseline) / 4) 0;
  border-bottom: 1px solid var(--layout);
}

<div>
  <div
    class="c0"
  >
    <p>
      my list
    </p>
    <button
      class="c1"
    >
      <svg />
    </button>
  </div>
  <ul
    class="c2"
  >
    <li>
      item1
    </li>
    <li>
      item2
    </li>
  </ul>
</div>
`;
